### DISCUSSION TITLE
How to manually trigger the auto generate typegen (eg: through on file save / a hotkey)?

### DISCUSSION THREAD

**Discussion by Norlandz** [statelyai/xstate-tools#408] [Q&A]
**_Updated Post_**

How to manually trigger the auto generate typegen (eg: through on file save / a hotkey)?

- Because::
  
  Currently, (seems), 
  Xstate will **only** trigger auto generate typegen, 
  if it sees `tsTypes: {} as import('./WebrtcConnectionStateMachine.typegen').Typegen0,`,
  
  This is **limited**. 
  Because I could use another approach 
  ```ts
  import { Typegen0 } from './WebrtcConnectionStateMachine.typegen';
  
  export const webrtcConnectionStateMachine = createMachine<
    WebrtcConnectionStateMachineContext,
    WebrtcConnectionStateMachineEvent,
    WebrtcConnectionStateMachineTypestate,
    WebrtcConnectionStateMachineService,
    Typegen0
  >(
  ```
  now Xstate **wont** trigger auto generate typegen.

---
**_Old Post_**

<strike>

- If I have this line `tsTypes: {} as import('./WebrtcConnectionStateMachine.typegen').Typegen0,`,
  then Xstate **will auto generate** typegen, 
  - Is this the only way?
  
- If I use `import { Typegen0 } from './WebrtcConnectionStateMachine.typegen';`, 
  then Xstate will **NOT** auto generate typegen.
  - Did I do something wrong?
  - ```ts
    import { Typegen0 } from './WebrtcConnectionStateMachine.typegen';
    
    export const webrtcConnectionStateMachine = createMachine<
      WebrtcConnectionStateMachineContext,
      WebrtcConnectionStateMachineEvent,
      WebrtcConnectionStateMachineTypestate,
      WebrtcConnectionStateMachineService,
      Typegen0
    >(
    ```
  
- (If I specify both,
  then error: `Generics cannot be specified using type parameters when using typegen. Use the schema property of the machine config instead.`)

</strike>

---
related issue:
https://github.com/statelyai/xstate-tools/issues/326

**Comment by Andarist** [comment]
Overall, you probably shouldn't use explicit generics with `xstate`. We can do **way more** for you if you embrace inference - and that will be even more true with the upcoming v5.

**Reply by Norlandz** [reply]
Maybe you misunderstood? 
`Typegen0` is indeed auto generated by Xstate. 

The problem is: 
Its only triggered when I put `tsTypes: {} as import('./WebrtcConnectionStateMachine.typegen').Typegen0`.
So I have to then manually comment out that line & re-import it using `import { Typegen0 } from './WebrtcConnectionStateMachine.typegen';`.

(Maybe I should change the question post 
from 
How to enable auto generate typegen properly in Xstate Vscode extension?
to 
How to manually trigger the auto generate typegen (eg: through on file save / a hotkey)?
)

**Reply by Andarist** [reply]
> Its only triggered when I put tsTypes: {} as import('./WebrtcConnectionStateMachine.typegen').Typegen0.

This is the way it works. For various reasons, we don't want you to use explicit type params between angle brackets.

**Reply by Norlandz** [reply]
@Andarist 

So
```ts
    schema: {
      context: {} as WebrtcConnectionStateMachineContext,
      events: {} as WebrtcConnectionStateMachineEvent,
      services: {} as WebrtcConnectionStateMachineService,
    },
    tsTypes: {} as import('./WebrtcConnectionStateMachine.typegen').Typegen0,
```
is better than
```ts
export const webrtcConnectionStateMachine = createMachine<
  WebrtcConnectionStateMachineContext,
  WebrtcConnectionStateMachineEvent,
  WebrtcConnectionStateMachineTypestate,
  WebrtcConnectionStateMachineService,
  Typegen0
>(
```
?

(forgot why I switched to the other way)
(also the Xstate way just feels hacky)

**Comment by Norlandz** [comment]
> Download and install the [VS Code extension (opens new window)](https://marketplace.visualstudio.com/items?itemName=statelyai.stately-vscode) OR install the [CLI](https://xstate.js.org/docs/packages/xstate-cli/) and run the `xstate typegen` command with the `--watch` flag. 
> 
> https://xstate.js.org/docs/guides/typescript.html?#typegen

This seems like another way? 
Or just the same behavior? 
<strike>
But is this triggered on file save?
Is there an easier one-button-click solution on the vscode extension, so I dont have to bother with CLI, at current stage?
</strike>

**Comment by Norlandz** [comment]
(close this as a duplicate of `#326` (after question changed))
